File: ts/src/wrappers/bedrock.ts
/**
 * AWS Bedrock Runtime Wrapper
 * Wraps @aws-sdk/client-bedrock-runtime client to auto-capture all Converse calls
 *
 * Supports:
 * - client.send(new ConverseCommand()) (non-streaming)
 * - client.send(new ConverseStreamCommand()) (streaming)
 * - All Bedrock foundation models (Claude, Llama, Titan, Mistral, etc.)
 */

import type { TraceData, RaindropRequestOptions, WithTraceId, InteractionContext, SpanData } from '../types.js';

/**
 * Safely parse JSON, returning the raw string if parsing fails.
 */
function safeJsonParse(str: string): unknown {
  try {
    return JSON.parse(str);
  } catch {
    return str;
  }
}

// Minimal type definitions for @aws-sdk/client-bedrock-runtime
// We use loose typing to avoid hard dependency on the AWS SDK

type BedrockClient = {
  send: (command: unknown) => Promise<unknown>;
};

type ConverseInput = {
  modelId: string;
  messages: Array<{
    role: 'user' | 'assistant';
    content: Array<{
      text?: string;
      image?: unknown;
      document?: unknown;
      toolUse?: { toolUseId: string; name: string; input: unknown };
      toolResult?: { toolUseId: string; content: unknown };
    }>;
  }>;
  system?: Array<{ text: string }>;
  inferenceConfig?: {
    maxTokens?: number;
    temperature?: number;
    topP?: number;
    stopSequences?: string[];
  };
  toolConfig?: {
    tools?: Array<{
      toolSpec?: {
        name: string;
        description?: string;
        inputSchema?: unknown;
      };
    }>;
  };
  [key: string]: unknown;
};

type ConverseOutput = {
  output?: {
    message?: {
      role: string;
      content: Array<{
        text?: string;
        toolUse?: { toolUseId: string; name: string; input: unknown };
      }>;
    };
  };
  stopReason?: string;
  usage?: {
    inputTokens: number;
    outputTokens: number;
    totalTokens: number;
  };
  metrics?: {
    latencyMs: number;
  };
};

type ConverseStreamOutput = {
  stream?: AsyncIterable<StreamEvent>;
};

type StreamEvent = {
  messageStart?: { role: string };
  contentBlockStart?: { contentBlockIndex: number; start?: { toolUse?: { toolUseId: string; name: string } } };
  contentBlockDelta?: { contentBlockIndex: number; delta?: { text?: string; toolUse?: { input?: string } } };
  contentBlockStop?: { contentBlockIndex: number };
  messageStop?: { stopReason: string };
  metadata?: { usage?: { inputTokens: number; outputTokens: number; totalTokens: number }; metrics?: { latencyMs: number } };
};

// Command type detection
interface CommandLike {
  input?: ConverseInput;
  constructor?: { name?: string };
}

export interface BedrockWrapperContext {
  generateTraceId: () => string;
  sendTrace: (trace: TraceData) => void;
  getUserId: () => string | undefined;
  getInteractionContext: () => InteractionContext | undefined;
  notifySpan: (span: SpanData) => void;
  debug: boolean;
}

/**
 * Creates a wrapped Bedrock client that auto-traces all Converse calls
 */
export function wrapBedrock<T extends BedrockClient>(
  client: T,
  context: BedrockWrapperContext
): T {
  const originalSend = client.send.bind(client);

  const wrappedSend = async (
    command: CommandLike,
    options?: { raindrop?: RaindropRequestOptions }
  ): Promise<unknown> => {
    const commandName = command.constructor?.name || '';

    // Only trace Converse commands
    if (commandName === 'ConverseCommand') {
      return handleConverse(originalSend, command, options, context);
    }

    if (commandName === 'ConverseStreamCommand') {
      return handleConverseStream(originalSend, command, options, context);
    }

    // Pass through other commands unchanged
    return originalSend(command);
  };

  // Create wrapped client
  const wrappedClient = Object.create(Object.getPrototypeOf(client));
  Object.assign(wrappedClient, client);
  wrappedClient.send = wrappedSend;

  return wrappedClient as T;
}

/**
 * Handle ConverseCommand (non-streaming)
 */
async function handleConverse(
  originalSend: (command: unknown) => Promise<unknown>,
  command: CommandLike,
  options: { raindrop?: RaindropRequestOptions } | undefined,
  context: BedrockWrapperContext
): Promise<WithTraceId<ConverseOutput>> {
  const input = command.input as ConverseInput;
  const traceId = options?.raindrop?.traceId || context.generateTraceId();
  const startTime = Date.now();
  const userId = options?.raindrop?.userId || context.getUserId();

  if (context.debug) {
    console.log('[raindrop] Bedrock Converse started:', traceId);
  }

  try {
    const response = await originalSend(command) as ConverseOutput;
    const endTime = Date.now();

    // Extract output
    const outputMessage = response.output?.message;
    const outputText = outputMessage?.content
      ?.filter(c => c.text)
      .map(c => c.text)
      .join('') || '';

    // Extract tool calls
    const toolCalls = outputMessage?.content
      ?.filter(c => c.toolUse)
      .map(c => ({
        id: c.toolUse!.toolUseId,
        name: c.toolUse!.name,
        arguments: c.toolUse!.input,
      })) || [];

    // Infer provider from model ID
    const provider = inferProvider(input.modelId);

    // Check if we're within an interaction
    const interaction = context.getInteractionContext();

    if (interaction) {
      const span: SpanData = {
        spanId: traceId,
        parentId: interaction.interactionId,
        name: `bedrock:${input.modelId}`,
        type: 'ai',
        startTime,
        endTime,
        latencyMs: endTime - startTime,
        input: input.messages,
        output: outputText,
        properties: {
          ...options?.raindrop?.properties,
          input_tokens: response.usage?.inputTokens,
          output_tokens: response.usage?.outputTokens,
          stop_reason: response.stopReason,
          tool_calls: toolCalls.length > 0 ? toolCalls : undefined,
        },
      };
      context.notifySpan(span);
      interaction.spans.push(span);
    } else {
      context.sendTrace({
        traceId,
        provider,
        model: input.modelId,
        input: input.messages,
        output: outputText,
        startTime,
        endTime,
        latencyMs: endTime - startTime,
        tokens: response.usage ? {
          input: response.usage.inputTokens,
          output: response.usage.outputTokens,
          total: response.usage.totalTokens,
        } : undefined,
        toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
        userId,
        conversationId: options?.raindrop?.conversationId,
        properties: {
          ...options?.raindrop?.properties,
          stop_reason: response.stopReason,
        },
      });
    }

    return Object.assign(response, { _traceId: traceId });
  } catch (error) {
    const endTime = Date.now();
    const interaction = context.getInteractionContext();
    const provider = inferProvider(input.modelId);

    if (interaction) {
      const span: SpanData = {
        spanId: traceId,
        parentId: interaction.interactionId,
        name: `bedrock:${input.modelId}`,
        type: 'ai',
        startTime,
        endTime,
        latencyMs: endTime - startTime,
        input: input.messages,
        error: error instanceof Error ? error.message : String(error),
      };
      context.notifySpan(span);
      interaction.spans.push(span);
    } else {
      context.sendTrace({
        traceId,
        provider,
        model: input.modelId,
        input: input.messages,
        startTime,
        endTime,
        latencyMs: endTime - startTime,
        userId,
        conversationId: options?.raindrop?.conversationId,
        properties: options?.raindrop?.properties,
        error: error instanceof Error ? error.message : String(error),
      });
    }

    throw error;
  }
}

/**
 * Handle ConverseStreamCommand (streaming)
 */
async function handleConverseStream(
  originalSend: (command: unknown) => Promise<unknown>,
  command: CommandLike,
  options: { raindrop?: RaindropRequestOptions } | undefined,
  context: BedrockWrapperContext
): Promise<WithTraceId<ConverseStreamOutput>> {
  const input = command.input as ConverseInput;
  const traceId = options?.raindrop?.traceId || context.generateTraceId();
  const startTime = Date.now();
  const userId = options?.raindrop?.userId || context.getUserId();

  if (context.debug) {
    console.log('[raindrop] Bedrock ConverseStream started:', traceId);
  }

  const response = await originalSend(command) as ConverseStreamOutput;

  if (!response.stream) {
    return Object.assign(response, { _traceId: traceId });
  }

  // Wrap the stream
  const wrappedStream = wrapStream(response.stream, {
    traceId,
    startTime,
    userId,
    conversationId: options?.raindrop?.conversationId,
    properties: options?.raindrop?.properties,
    modelId: input.modelId,
    input: input.messages,
    context,
  });

  return Object.assign({ stream: wrappedStream }, { _traceId: traceId });
}

/**
 * Wrap a stream to capture events and send trace on completion
 */
function wrapStream(
  stream: AsyncIterable<StreamEvent>,
  options: {
    traceId: string;
    startTime: number;
    userId?: string;
    conversationId?: string;
    properties?: Record<string, unknown>;
    modelId: string;
    input: unknown;
    context: BedrockWrapperContext;
  }
): AsyncIterable<StreamEvent> {
  const interaction = options.context.getInteractionContext();
  const collectedText: string[] = [];
  const toolCalls: Map<number, { id: string; name: string; arguments: string }> = new Map();
  let usage: { inputTokens: number; outputTokens: number; totalTokens: number } | undefined;
  let stopReason: string | undefined;

  return {
    [Symbol.asyncIterator]: async function* () {
      try {
        for await (const event of stream) {
          // Collect text deltas
          if (event.contentBlockDelta?.delta?.text) {
            collectedText.push(event.contentBlockDelta.delta.text);
          }

          // Collect tool use start
          if (event.contentBlockStart?.start?.toolUse) {
            const idx = event.contentBlockStart.contentBlockIndex;
            toolCalls.set(idx, {
              id: event.contentBlockStart.start.toolUse.toolUseId,
              name: event.contentBlockStart.start.toolUse.name,
              arguments: '',
            });
          }

          // Collect tool use input deltas
          if (event.contentBlockDelta?.delta?.toolUse?.input) {
            const idx = event.contentBlockDelta.contentBlockIndex;
            const existing = toolCalls.get(idx);
            if (existing) {
              existing.arguments += event.contentBlockDelta.delta.toolUse.input;
            }
          }

          // Collect stop reason
          if (event.messageStop?.stopReason) {
            stopReason = event.messageStop.stopReason;
          }

          // Collect usage from metadata
          if (event.metadata?.usage) {
            usage = event.metadata.usage;
          }

          yield event;
        }

        // Stream complete - send trace
        const endTime = Date.now();
        const output = collectedText.join('');
        const provider = inferProvider(options.modelId);

        const parsedToolCalls = Array.from(toolCalls.values()).map(tc => ({
          id: tc.id,
          name: tc.name,
          arguments: tc.arguments ? safeJsonParse(tc.arguments) : {},
        }));

        if (interaction) {
          const span: SpanData = {
            spanId: options.traceId,
            parentId: interaction.interactionId,
            name: `bedrock:${options.modelId}`,
            type: 'ai',
            startTime: options.startTime,
            endTime,
            latencyMs: endTime - options.startTime,
            input: options.input,
            output,
            properties: {
              ...options.properties,
              input_tokens: usage?.inputTokens,
              output_tokens: usage?.outputTokens,
              stop_reason: stopReason,
              tool_calls: parsedToolCalls.length > 0 ? parsedToolCalls : undefined,
            },
          };
          options.context.notifySpan(span);
          interaction.spans.push(span);
        } else {
          options.context.sendTrace({
            traceId: options.traceId,
            provider,
            model: options.modelId,
            input: options.input,
            output,
            startTime: options.startTime,
            endTime,
            latencyMs: endTime - options.startTime,
            tokens: usage ? {
              input: usage.inputTokens,
              output: usage.outputTokens,
              total: usage.totalTokens,
            } : undefined,
            toolCalls: parsedToolCalls.length > 0 ? parsedToolCalls : undefined,
            userId: options.userId,
            conversationId: options.conversationId,
            properties: {
              ...options.properties,
              stop_reason: stopReason,
            },
          });
        }
      } catch (error) {
        const endTime = Date.now();
        const provider = inferProvider(options.modelId);

        if (interaction) {
          const span: SpanData = {
            spanId: options.traceId,
            parentId: interaction.interactionId,
            name: `bedrock:${options.modelId}`,
            type: 'ai',
            startTime: options.startTime,
            endTime,
            latencyMs: endTime - options.startTime,
            input: options.input,
            error: error instanceof Error ? error.message : String(error),
          };
          options.context.notifySpan(span);
          interaction.spans.push(span);
        } else {
          options.context.sendTrace({
            traceId: options.traceId,
            provider,
            model: options.modelId,
            input: options.input,
            startTime: options.startTime,
            endTime,
            latencyMs: endTime - options.startTime,
            userId: options.userId,
            conversationId: options.conversationId,
            properties: options.properties,
            error: error instanceof Error ? error.message : String(error),
          });
        }

        throw error;
      }
    },
  };
}

/**
 * Infer provider from Bedrock model ID
 */
function inferProvider(modelId: string): string {
  const id = modelId.toLowerCase();

  if (id.includes('anthropic') || id.includes('claude')) return 'anthropic';
  if (id.includes('amazon') || id.includes('titan') || id.includes('nova')) return 'amazon';
  if (id.includes('meta') || id.includes('llama')) return 'meta';
  if (id.includes('mistral') || id.includes('mixtral')) return 'mistral';
  if (id.includes('cohere') || id.includes('command')) return 'cohere';
  if (id.includes('ai21') || id.includes('jamba') || id.includes('jurassic')) return 'ai21';
  if (id.includes('stability') || id.includes('stable')) return 'stability';

  return 'bedrock';
}

File: python/src/rd_mini/wrappers/bedrock.py
"""
AWS Bedrock Runtime Wrapper
Wraps boto3 bedrock-runtime client to auto-capture all Converse calls

Supports:
- client.converse() (non-streaming)
- client.converse_stream() (streaming)
- All Bedrock foundation models (Claude, Llama, Titan, Mistral, etc.)
"""

from __future__ import annotations

import json
import time
from typing import TYPE_CHECKING, Any, AsyncIterator, Callable, Iterator

from rd_mini.types import InteractionContext, SpanData, TraceData

if TYPE_CHECKING:
    pass


class WrapperContext:
    """Context passed to wrappers."""

    def __init__(
        self,
        generate_trace_id: Callable[[], str],
        send_trace: Callable[[TraceData], None],
        get_user_id: Callable[[], str | None],
        get_interaction_context: Callable[[], InteractionContext | None],
        debug: bool,
    ):
        self.generate_trace_id = generate_trace_id
        self.send_trace = send_trace
        self.get_user_id = get_user_id
        self.get_interaction_context = get_interaction_context
        self.debug = debug


def safe_json_loads(s: str) -> Any:
    """Safely parse JSON, returning the raw string if parsing fails."""
    try:
        return json.loads(s)
    except json.JSONDecodeError:
        return s


def _infer_provider(model_id: str) -> str:
    """Infer provider from Bedrock model ID."""
    id_lower = model_id.lower()

    if "anthropic" in id_lower or "claude" in id_lower:
        return "anthropic"
    if "amazon" in id_lower or "titan" in id_lower or "nova" in id_lower:
        return "amazon"
    if "meta" in id_lower or "llama" in id_lower:
        return "meta"
    if "mistral" in id_lower or "mixtral" in id_lower:
        return "mistral"
    if "cohere" in id_lower or "command" in id_lower:
        return "cohere"
    if "ai21" in id_lower or "jamba" in id_lower or "jurassic" in id_lower:
        return "ai21"
    if "stability" in id_lower or "stable" in id_lower:
        return "stability"

    return "bedrock"


class WrappedBedrockClient:
    """Wrapped Bedrock client that traces converse calls."""

    def __init__(self, client: Any, context: WrapperContext):
        self._client = client
        self._context = context

    def converse(
        self,
        *args: Any,
        raindrop: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Any:
        """Converse with automatic tracing."""
        trace_id = (raindrop or {}).get("trace_id") or self._context.generate_trace_id()
        start_time = time.time()
        user_id = (raindrop or {}).get("user_id") or self._context.get_user_id()
        conversation_id = (raindrop or {}).get("conversation_id")
        properties = (raindrop or {}).get("properties", {})

        model_id = kwargs.get("modelId", "unknown")
        messages = kwargs.get("messages", [])

        if self._context.debug:
            print(f"[raindrop] Bedrock converse started: {trace_id}")

        try:
            response = self._client.converse(*args, **kwargs)
            end_time = time.time()

            # Extract output
            output_message = response.get("output", {}).get("message", {})
            output_content = output_message.get("content", [])

            output_text = ""
            tool_calls = []

            for block in output_content:
                if "text" in block:
                    output_text += block["text"]
                if "toolUse" in block:
                    tool_use = block["toolUse"]
                    tool_calls.append({
                        "id": tool_use.get("toolUseId", ""),
                        "name": tool_use.get("name", ""),
                        "arguments": tool_use.get("input", {}),
                    })

            # Extract usage
            usage = response.get("usage", {})
            tokens = None
            if usage:
                tokens = {
                    "input": usage.get("inputTokens", 0),
                    "output": usage.get("outputTokens", 0),
                    "total": usage.get("totalTokens", 0),
                }

            stop_reason = response.get("stopReason")
            provider = _infer_provider(model_id)

            # Check for interaction context
            interaction = self._context.get_interaction_context()

            if interaction:
                span = SpanData(
                    span_id=trace_id,
                    parent_id=interaction.interaction_id,
                    name=f"bedrock:{model_id}",
                    type="ai",
                    start_time=start_time,
                    end_time=end_time,
                    latency_ms=int((end_time - start_time) * 1000),
                    input=messages,
                    output=output_text,
                    properties={
                        **properties,
                        "input_tokens": tokens["input"] if tokens else None,
                        "output_tokens": tokens["output"] if tokens else None,
                        "stop_reason": stop_reason,
                        "tool_calls": tool_calls if tool_calls else None,
                    },
                )
                interaction.spans.append(span)
            else:
                self._context.send_trace(
                    TraceData(
                        trace_id=trace_id,
                        provider=provider,
                        model=model_id,
                        input=messages,
                        output=output_text,
                        start_time=start_time,
                        end_time=end_time,
                        latency_ms=int((end_time - start_time) * 1000),
                        tokens=tokens,
                        tool_calls=tool_calls if tool_calls else None,
                        user_id=user_id,
                        conversation_id=conversation_id,
                        properties={
                            **properties,
                            "stop_reason": stop_reason,
                        },
                    )
                )

            # Attach trace_id to response
            response["_trace_id"] = trace_id
            return response

        except Exception as e:
            end_time = time.time()
            interaction = self._context.get_interaction_context()
            provider = _infer_provider(model_id)

            if interaction:
                span = SpanData(
                    span_id=trace_id,
                    parent_id=interaction.interaction_id,
                    name=f"bedrock:{model_id}",
                    type="ai",
                    start_time=start_time,
                    end_time=end_time,
                    latency_ms=int((end_time - start_time) * 1000),
                    input=messages,
                    error=str(e),
                )
                interaction.spans.append(span)
            else:
                self._context.send_trace(
                    TraceData(
                        trace_id=trace_id,
                        provider=provider,
                        model=model_id,
                        input=messages,
                        start_time=start_time,
                        end_time=end_time,
                        latency_ms=int((end_time - start_time) * 1000),
                        user_id=user_id,
                        conversation_id=conversation_id,
                        properties=properties,
                        error=str(e),
                    )
                )
            raise

    def converse_stream(
        self,
        *args: Any,
        raindrop: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Any:
        """Converse with streaming and automatic tracing."""
        trace_id = (raindrop or {}).get("trace_id") or self._context.generate_trace_id()
        start_time = time.time()
        user_id = (raindrop or {}).get("user_id") or self._context.get_user_id()
        conversation_id = (raindrop or {}).get("conversation_id")
        properties = (raindrop or {}).get("properties", {})

        model_id = kwargs.get("modelId", "unknown")
        messages = kwargs.get("messages", [])

        if self._context.debug:
            print(f"[raindrop] Bedrock converse_stream started: {trace_id}")

        response = self._client.converse_stream(*args, **kwargs)

        # The response contains a 'stream' key with the event stream
        stream = response.get("stream")
        if not stream:
            response["_trace_id"] = trace_id
            return response

        # Wrap the stream
        wrapped_stream = TracedBedrockStream(
            stream=stream,
            trace_id=trace_id,
            start_time=start_time,
            user_id=user_id,
            conversation_id=conversation_id,
            properties=properties,
            model_id=model_id,
            messages=messages,
            context=self._context,
        )

        # Return response with wrapped stream
        result = dict(response)
        result["stream"] = wrapped_stream
        result["_trace_id"] = trace_id
        return result

    def __getattr__(self, name: str) -> Any:
        """Forward other attributes to original client."""
        return getattr(self._client, name)


class TracedBedrockStream:
    """Wrapper around Bedrock stream that traces on completion."""

    def __init__(
        self,
        stream: Any,
        trace_id: str,
        start_time: float,
        user_id: str | None,
        conversation_id: str | None,
        properties: dict[str, Any],
        model_id: str,
        messages: list[Any],
        context: WrapperContext,
    ):
        self._stream = stream
        self.__trace_id = trace_id
        self._start_time = start_time
        self._user_id = user_id
        self._conversation_id = conversation_id
        self._properties = properties
        self._model_id = model_id
        self._messages = messages
        self._context = context
        self._collected_text: list[str] = []
        self._tool_calls: dict[int, dict[str, Any]] = {}
        self._usage: dict[str, int] | None = None
        self._stop_reason: str | None = None
        self._interaction = context.get_interaction_context()

    @property
    def _trace_id(self) -> str:
        return self.__trace_id

    def __iter__(self) -> Iterator[Any]:
        try:
            for event in self._stream:
                self._process_event(event)
                yield event

            # Stream complete - send trace
            self._finalize()

        except Exception as e:
            self._finalize(error=str(e))
            raise

    async def __aiter__(self) -> AsyncIterator[Any]:
        try:
            async for event in self._stream:
                self._process_event(event)
                yield event

            # Stream complete - send trace
            self._finalize()

        except Exception as e:
            self._finalize(error=str(e))
            raise

    def _process_event(self, event: dict[str, Any]) -> None:
        """Process a stream event and collect data."""
        # Content block delta - text
        if "contentBlockDelta" in event:
            delta = event["contentBlockDelta"].get("delta", {})
            if "text" in delta:
                self._collected_text.append(delta["text"])
            # Tool use input delta
            if "toolUse" in delta and "input" in delta["toolUse"]:
                idx = event["contentBlockDelta"].get("contentBlockIndex", 0)
                if idx in self._tool_calls:
                    self._tool_calls[idx]["arguments"] += delta["toolUse"]["input"]

        # Content block start - tool use
        if "contentBlockStart" in event:
            start = event["contentBlockStart"].get("start", {})
            if "toolUse" in start:
                idx = event["contentBlockStart"].get("contentBlockIndex", 0)
                self._tool_calls[idx] = {
                    "id": start["toolUse"].get("toolUseId", ""),
                    "name": start["toolUse"].get("name", ""),
                    "arguments": "",
                }

        # Message stop
        if "messageStop" in event:
            self._stop_reason = event["messageStop"].get("stopReason")

        # Metadata with usage
        if "metadata" in event:
            usage = event["metadata"].get("usage", {})
            if usage:
                self._usage = {
                    "input": usage.get("inputTokens", 0),
                    "output": usage.get("outputTokens", 0),
                    "total": usage.get("totalTokens", 0),
                }

    def _finalize(self, error: str | None = None) -> None:
        """Send trace on stream completion."""
        end_time = time.time()
        output = "".join(self._collected_text)
        provider = _infer_provider(self._model_id)

        # Parse tool call arguments
        parsed_tool_calls = []
        for tc in self._tool_calls.values():
            parsed_tool_calls.append({
                "id": tc["id"],
                "name": tc["name"],
                "arguments": safe_json_loads(tc["arguments"]) if tc["arguments"] else {},
            })

        if self._interaction:
            span = SpanData(
                span_id=self._trace_id,
                parent_id=self._interaction.interaction_id,
                name=f"bedrock:{self._model_id}",
                type="ai",
                start_time=self._start_time,
                end_time=end_time,
                latency_ms=int((end_time - self._start_time) * 1000),
                input=self._messages,
                output=output if not error else None,
                error=error,
                properties={
                    **self._properties,
                    "input_tokens": self._usage["input"] if self._usage else None,
                    "output_tokens": self._usage["output"] if self._usage else None,
                    "stop_reason": self._stop_reason,
                    "tool_calls": parsed_tool_calls if parsed_tool_calls else None,
                },
            )
            self._interaction.spans.append(span)
        else:
            self._context.send_trace(
                TraceData(
                    trace_id=self._trace_id,
                    provider=provider,
                    model=self._model_id,
                    input=self._messages,
                    output=output if not error else None,
                    start_time=self._start_time,
                    end_time=end_time,
                    latency_ms=int((end_time - self._start_time) * 1000),
                    tokens=self._usage,
                    tool_calls=parsed_tool_calls if parsed_tool_calls else None,
                    user_id=self._user_id,
                    conversation_id=self._conversation_id,
                    properties={
                        **self._properties,
                        "stop_reason": self._stop_reason,
                    },
                    error=error,
                )
            )


def wrap_bedrock(client: Any, context: WrapperContext) -> WrappedBedrockClient:
    """Wrap a Bedrock client for automatic tracing."""
    return WrappedBedrockClient(client, context)
